py_ecc/fields/optimized_field_elements.py

```diff
@@ -0,0 +1,395 @@
+                .format(type(other))
+            )
+
+        return FQ(
+            self.n * prime_field_inv(on, self.field_modulus) % self.field_modulus,
+            self.curve_name
+        )
+
+    def __truediv__(self, other: IntOrFQ) -> "FQ":
+        return self.__div__(other)
+
+    def __rdiv__(self, other: IntOrFQ) -> "FQ":
+        if isinstance(other, FQ):
+            on = other.n
+        elif isinstance(other, int):
+            on = other
+        else:
+            raise ValueError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(other))
+            )
+
+        return FQ(
+            prime_field_inv(self.n, self.field_modulus) * on % self.field_modulus,
+            self.curve_name
+        )
+
+    def __rtruediv__(self, other: IntOrFQ) -> "FQ":
+        return self.__rdiv__(other)
+
+    def __pow__(self, other: int) -> "FQ":
+        if other == 0:
+            return FQ(1, self.curve_name)
+        elif other == 1:
+            return FQ(self.n, self.curve_name)
+        elif other % 2 == 0:
+            return (self * self) ** (other // 2)
+        else:
+            return ((self * self) ** int(other // 2)) * self
+
+    def __eq__(self, other: IntOrFQ) -> bool:  # type:ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        if isinstance(other, FQ):
+            return self.n == other.n
+        elif isinstance(other, int):
+            return self.n == other
+        else:
+            raise ValueError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(other))
+            )
+
+    def __ne__(self, other: IntOrFQ) -> bool:    # type:ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        return not self == other
+
+    def __neg__(self) -> "FQ":
+        return FQ(-self.n, self.curve_name)
+
+    def __repr__(self) -> str:
+        return repr(self.n)
+
+    def __int__(self) -> int:
+        return self.n
+
+    @classmethod
+    def one(cls, curve_name: str) -> "FQ":
+        return cls(1, curve_name)
+
+    @classmethod
+    def zero(cls, curve_name: str) -> "FQ":
+        return cls(0, curve_name)
+
+
+class FQP(object):
+    """
+    A class for elements in polynomial extension fields
+    """
+    degree = 0  # type: int
+    mc_tuples = None  # type: List[Tuple[int, int]]
+    curve_name = None
+    field_modulus = None
+
+    def __init__(self,
+                 coeffs: Sequence[IntOrFQ],
+                 curve_name: str,
+                 modulus_coeffs: Sequence[IntOrFQ]=None) -> None:
+        """
+        curve_name can be either 'bn128' or 'bls12_381'
+        This is needed to obtain field_modulus, FQ2_MODULUS_COEFFS
+        and FQ12_MODULUS_COEFFS from the curve properties
+        """
+        self.curve_name = curve_name
+        self.field_modulus = field_properties[curve_name]["field_modulus"]
```

Comment:
Again, I quite appreciate it that you kept it as similar as possible to the old `FQP`, for the first review.

*suuper nitpicky* but I'd be in favor of reordering the `init` lines so that they match up with the `FQP` in the non-optimized version (at least, where the lines are the same).

If you don't feel like it, no worries, we can add an issue and someone can come back later to clean it up (maybe a `Good First Issue`).